python/lsst/pipe/base/quantum_provenance_graph.py

```diff
@@ -0,0 +1,961 @@
+    speeds up comparisons.
+    """
+
+    is_task: ClassVar[Literal[True]] = True
+    """Whether this node represents a quantum rather
+    than a dataset (always `True`).
+    """
+
+
+class DatasetKey(NamedTuple):
+    """Identifier type for dataset keys in a `QuantumProvenanceGraph`."""
+
+    parent_dataset_type_name: str
+    """Name of the dataset type (never a component)."""
+
+    data_id_values: tuple[DataIdValue, ...]
+    """Data ID values of the dataset.
+
+    Note that keys are fixed given `parent_dataset_type_name`, so using only
+    the values here speeds up comparisons.
+    """
+
+    is_task: ClassVar[Literal[False]] = False
+    """Whether this node represents a quantum rather
+    than a dataset (always `False`).
+    """
+
+    is_prerequisite: ClassVar[Literal[False]] = False
+    """Whether this node is a prerequisite to another node (also always
+    `False`).
+    """
+
+
+class PrerequisiteDatasetKey(NamedTuple):
+    """Identifier type for prerequisite dataset keys in a
+    `QuantumProvenanceGraph`.
+
+    Unlike regular datasets, prerequisites are not actually required to come
+    from a find-first search of `input_collections`, so we don't want to
+    assume that the same data ID implies the same dataset.  Happily we also
+    don't need to search for them by data ID in the graph, so we can use the
+    dataset ID (UUID) instead.
+    """
+
+    parent_dataset_type_name: str
+    """Name of the dataset type (never a component)."""
+
+    dataset_id_bytes: bytes
+    """Dataset ID (UUID) as raw bytes."""
+
+    is_task: ClassVar[Literal[False]] = False
+    """Whether this node represents a quantum rather
+    than a dataset (always `False`).
+    """
+
+    is_prerequisite: ClassVar[Literal[True]] = True
+    """Whether this node is a prerequisite to another node (always `True`).
+    """
+
+
+QuantumRunStatus: TypeAlias = Literal["failed", "successful", "logs_missing", "blocked", "metadata_missing"]
+
+
+class QuantumRun(pydantic.BaseModel):
+    """Information about a quantum in a given run collection."""
+
+    id: uuid.UUID
+    """The quantum graph node ID associated with the dataId in a specific run.
+    """
+
+    status: QuantumRunStatus = "metadata_missing"
+    """The status of the quantum in that run.
+    """
+
+
+QuantumInfoStatus: TypeAlias = Literal["successful", "wonky", "blocked", "not_attempted", "failed"]
+
+
+class QuantumInfo(TypedDict):
+    """Information about a quantum across all run collections.
+
+    Used to annotate the networkx node dictionary.
+    """
+
+    data_id: DataCoordinate
+    """The data_id of the quantum.
+    """
+
+    runs: dict[str, QuantumRun]
+    """All run collections associated with the quantum.
+    """
+
+    status: QuantumInfoStatus
+    """The overall status of the quantum. Note that it is impossible to exit a
+    wonky state.
+    """
+
+    recovered: bool
+    """The quantum was originally not successful but was ultimately successful.
+    """
```

Comment:
Changing status to recovered turned out to mean a lot of code changes (because need to combine recovered with successfule) elsewhere.  So ok with revisiting this later if it comes up again.